### ISSUE TITLE
[DR] Near-axis magnetic field calculation from efit files

### ISSUE THREAD

**Issue by Maxwell-Rosen** [ammarhakim/gkylzero#439]
This change concerns the magnetic field calculation on axis when reading efit files. Before this change, we had issues because we were not treating the axis extra specially. The field equations are as follows.

$$B_R = \frac{1}{R} \frac{\partial \psi}{\partial Z} = \frac{\partial}{\partial Z} \left( \frac{\psi}{R} \right),$$

$$B_Z = - \frac{1}{R} \frac{\partial \psi}{\partial R} = - \frac{\partial}{\partial R} \left( \frac{\psi}{R} \right) - \frac{\psi}{R^2}.$$

The last equality is what Gkeyll does under the hood. The workflow goes that from the efit file, psi is read in. This is defined everywhere, including at $R=0$. Then the quantities are computed of $\psi/R$ and $\psi/R^2$. These make taking the derivatives easier since the part of the code that calculates the magnetic field does not know $R$. This breaks down at $R=0$ because we have a divide-by-zero error. Instead, we should take the limit of $B$ as $R \to 0$ and use L’Hopital’s rule to evaluate the limits.

$$B_R = \lim_{R \to 0} \frac{1}{R} \frac{\partial \psi}{\partial Z}  = \lim_{R \to 0} \frac{\partial^2 \psi}{\partial Z \partial R},$$

$$B_Z = - \lim_{R \to 0} \frac{1}{R} \frac{\partial \psi}{\partial R} = - \lim_{R \to 0} \frac{\partial^2 \psi}{\partial R^2}.$$

The modification activates special kernels to compute these second derivatives instead for when the values of B are returned as NaN due to the divide by zero error in calculating $\psi / R$ from the efit file. Using these near-axis kernels for the nearest index to R=0 is not necessarily enough because the computational grid does not know R directly, mainly when calculating the magnetic field. We will treat the cells extra specially using the near-axis kernels only after the far-axis kernels fail, which always results in a NaN in the B field. These NaNs in the B field occur due to NaNs in the stencil of $\psi/R$, but there are so many components that go into this stencil that it is difficult to check. The NaNs accumulate in the magnetic field's magnitude, which can be checked for NaNs in a single line of code.

**Comment by ammarhakim** [comment]
You can't check for NaNs in Gkeyll at all. This is impossible. Also, it is a terrible idea to check for NaNs (even if you can) to activate "special kernels". Please rethink this issue properly and figure out a robust way of computing the geometry as we get close to the axis.

**Comment by Maxwell-Rosen** [comment]
We do not have a value of R when the magnetic field is computed. The code pre-computes $\psi/R$ and $\psi/R^2$ when the geometry is read and then these quantities are used in the magnetic field calculation. Because these quantities are projected onto a DG basis at the computational nodes, we loose the information about the efit file's grid. We need to know if we are in the cell closest to $R=0$ on the efit file grid. Perhaps we can store another array to keep track of this.

**Comment by akashukla** [comment]
We don't need to check for Nans or anything like that. When we call the kernel, we know the value of R at the cell boundaries. We can call the special kernel when we are at the axis.

**Comment by Maxwell-Rosen** [comment]
Right. We do know the value of R, but we don't know which cell from the efit file this is in. We can store the efit file index for r in a gkyl array and check if this index is less than 1 anywhere to compute the special kernels

**Comment by JunoRavin** [comment]
I am very confused why you need the index. You just need to know the value of R. You know the value of R inside the kernel. You only call this kernel once. Just check inside the kernel if R is at the desired tolerance for switching to a different means of computing B (just like how in the kernel I sent you, I know the value of the density and I just tell it to use Lax fluxes if the density is small).

**Comment by Maxwell-Rosen** [comment]
I get what you mean. We can define a tolerance elsewhere for R at the boundary. One issue is that this tolerance is not that small and it does depend on what efit file you are using

**Comment by Maxwell-Rosen** [comment]
We will need a different tolerance for each different efit file depending on their resolution. Each will go near the axis, but to different extents. We can define something universal, but not efit files are the same. Since this is really just a mirror thing, I don't mind this. After some testing, I see how it would work and it's straightforward, but it's not universal therefore not robust

**Comment by ammarhakim** [comment]
This should not be dependent on efit file but delta R only. So you can choose a tolerance to be a small fraction of delta R. But, in any case, the problem here is still not solved. How will you ensure that Bz at R=0 is continuous?

**Comment by Maxwell-Rosen** [comment]
What is delta R? Is this the cell size in the efit file? Then this is dependent on efit file resolution. Is there some other issue with continuity other than ensuring that 
$$\frac{\partial B_Z}{\partial R} \bigg\rvert_{R=0} = 0$$ ?

**Comment by Maxwell-Rosen** [comment]
I did some testing and it seems like the efit cell size is already in this part of the code through the physical RZ grid. We can read the cell sizes there and determine if the special kernels need to be called from this. I checked and this is the correct logic for high resolution grids and seems to be placed along the correct boundary.

https://github.com/ammarhakim/gkylzero/blob/d83f7a0e25ce04e506e2ff30203fefb1eac01fca/zero/calc_bmag.c#L180

![image](https://github.com/user-attachments/assets/685c9d35-7e0c-4a5a-9b89-fc9e58703f77)
